astropy/io/fits/fitsrec.py

```diff
@@ -92,9 +92,10 @@ def __setitem__(self, key, value):
             if indx < self.start or indx > self.end - 1:
                 raise KeyError(f"Key '{key}' does not exist.")
         elif isinstance(key, slice):
-            for indx in range(slice.start, slice.stop, slice.step):
-                indx = self._get_indx(indx)
-                self.array.field(indx)[self.row] = value
+            for indx, val in zip(range(*key.indices(self.array._nfields)), value):
```

Comment:
Let's avoid dropping overflowing `value`s silently
```suggestion
            for indx, val in zip(range(*key.indices(self.array._nfields)), value, strict=True):
```